# Sommaire :
- <a href="#C1">Import des librairies</a>
- <a href="#C2">Import des données</a>
- <a href="#C3">Lecture des données & suppression granularités inutiles</a>
- <a href="#C4">Correspondance WaterServices et Mortality</a>
- <a href="#C5">Correspondance df et PoliticalStability</a>
- <a href="#C6">Correspondance df & Population</a>
- <a href="#C7">Correspondance df & RegionCountry</a>
- <a href="#C8">Fusion des datasets après nettoyage</a>
    - <a href="#C9"> Fusion WaterServices & Mortality </a>
    - <a href="#C10"> Fusion df & Political Stability </a>
    - <a href="#C11"> Fusion df1 & Population </a>
    - <a href="#C12"> Fusion df2 & Population </a>
- <a href="#C14.2">Ajout de la colonne Urban population rate</a>
- <a href="#C14.5"> Ajout des densités de Population </a>
    - <a href="#C14.51"> Ajout des densités de Population par Pays</a>
    - <a href="#C14.52"> Ajout des densités de Population par Région</a>
    - <a href="#C14.53"> Ajout de la densité de Population niveau Monde</a>
- <a href="#C15"> Export du .csv </a>


# <a name="C1"> Import des librairies </a>

In [318]:
import pandas as pd
pd.options.display.max_columns = None # pour afficher toutes les colonnes
pd.options.display.max_rows = None # pour afficher toutes les lignes

import numpy as np

from scipy import stats

from datetime import datetime

import matplotlib as mpl
import matplotlib.pyplot as plt
import matplotlib.ticker as plticker

import seaborn as sns

import statsmodels.api as sm
import pylab

# <a name="C2"> Import des données </a>

Données d'OpenClassrooms et de la Banque mondiale pour les surfaces des pays

In [319]:
Population = pd.read_csv('Population.csv')
BasicAndSafelyManagedDrinkingWaterServices = pd.read_csv('BasicAndSafelyManagedDrinkingWaterServices.csv')
PoliticalStability = pd.read_csv('PoliticalStability.csv')
MortalityRateAttributedToWater = pd.read_csv('MortalityRateAttributedToWater.csv')
RegionCountry = pd.read_csv('RegionCountry.csv')

# <a name="C3"> Lecture des données & suppression granularités inutiles </a>

## BasicAndSafelyManagedDrinkingWaterServices

In [320]:
BasicAndSafelyManagedDrinkingWaterServices.head()

Year      Country Granularity  \
0  2000  Afghanistan       Rural   
1  2000  Afghanistan       Total   
2  2000  Afghanistan       Urban   
3  2000      Albania       Rural   
4  2000      Albania       Total   

   Population using at least basic drinking-water services (%)  \
0                                           21.61913             
1                                           27.77190             
2                                           49.48745             
3                                           81.78472             
4                                           87.86662             

   Population using safely managed drinking-water services (%)  
0                                                NaN            
1                                                NaN            
2                                                NaN            
3                                                NaN            
4                                           49.29324

In [321]:
len(BasicAndSafelyManagedDrinkingWaterServices)

10476

In [322]:
BasicAndSafelyManagedDrinkingWaterServices['Granularity'].value_counts()

Urban    3492
Rural    3492
Total    3492
Name: Granularity, dtype: int64

## MortalityRateAttributedToWater

In [323]:
MortalityRateAttributedToWater.head()

Year      Country Granularity  \
0  2016  Afghanistan      Female   
1  2016  Afghanistan        Male   
2  2016  Afghanistan       Total   
3  2016      Albania      Female   
4  2016      Albania        Male   

   Mortality rate attributed to exposure to unsafe WASH services  WASH deaths  
0                                           15.31193                      NaN  
1                                           12.61297                      NaN  
2                                           13.92067                 4824.353  
3                                            0.12552                      NaN  
4                                            0.20650                      NaN

In [324]:
len(MortalityRateAttributedToWater)

549

In [325]:
MortalityRateAttributedToWater['Granularity'].value_counts()

Total     183
Male      183
Female    183
Name: Granularity, dtype: int64

Suppression des granularités Male/Female :

In [326]:
MortalityRateAttributedToWater.drop\
(MortalityRateAttributedToWater[MortalityRateAttributedToWater['Granularity'] == 'Male'].index,\
 inplace=True)
MortalityRateAttributedToWater.drop\
(MortalityRateAttributedToWater[MortalityRateAttributedToWater['Granularity'] == 'Female'].index,\
 inplace=True)

MortalityRateAttributedToWater['Granularity'].value_counts() # vérification

Total    183
Name: Granularity, dtype: int64

## PoliticalStability

In [327]:
PoliticalStability.head()

Country  Year  Political_Stability Granularity
0  Afghanistan  2000                -2.44       Total
1  Afghanistan  2002                -2.04       Total
2  Afghanistan  2003                -2.20       Total
3  Afghanistan  2004                -2.30       Total
4  Afghanistan  2005                -2.07       Total

In [328]:
len(PoliticalStability)

3526

In [329]:
PoliticalStability['Granularity'].value_counts()

Total    3526
Name: Granularity, dtype: int64

## Population

In [330]:
Population.head()

Country Granularity  Year  Population
0  Afghanistan       Total  2000   20779.953
1  Afghanistan        Male  2000   10689.508
2  Afghanistan      Female  2000   10090.449
3  Afghanistan       Rural  2000   15657.474
4  Afghanistan       Urban  2000    4436.282

In [331]:
len(Population)

20914

In [332]:
Population['Granularity'].value_counts()

Total     4430
Urban     4414
Rural     4414
Male      3828
Female    3828
Name: Granularity, dtype: int64

Suppression des granularités Male/Female :

In [333]:
Population.drop\
(Population[Population['Granularity'] == 'Male'].index,inplace=True)
Population.drop\
(Population[Population['Granularity'] == 'Female'].index,inplace=True)

Population['Granularity'].value_counts() # vérification

Total    4430
Urban    4414
Rural    4414
Name: Granularity, dtype: int64

## RegionCountry

In [334]:
RegionCountry.head()

REGION (DISPLAY) COUNTRY (DISPLAY)
0           Europe           Albania
1           Europe           Andorra
2           Europe           Armenia
3  Western Pacific         Australia
4           Europe           Austria

In [335]:
len(RegionCountry)

194

In [336]:
RegionCountry['REGION (DISPLAY)'].value_counts()

Europe                   53
Africa                   47
Americas                 35
Western Pacific          27
Eastern Mediterranean    21
South-East Asia          11
Name: REGION (DISPLAY), dtype: int64

### Changements RegionDisplay

In [337]:
# Pays dans Eastern Meditteranean placés en Afrique
RegionCountry['REGION (DISPLAY)'].mask(RegionCountry['COUNTRY (DISPLAY)'] == 'Morocco', 'Africa' , inplace=True)
RegionCountry['REGION (DISPLAY)'].mask(RegionCountry['COUNTRY (DISPLAY)'] == 'Tunisia', 'Africa' , inplace=True)
RegionCountry['REGION (DISPLAY)'].mask(RegionCountry['COUNTRY (DISPLAY)'] == 'Egypt', 'Africa' , inplace=True)
RegionCountry['REGION (DISPLAY)'].mask(RegionCountry['COUNTRY (DISPLAY)'] == 'Lybia', 'Africa' , inplace=True)
RegionCountry['REGION (DISPLAY)'].mask(RegionCountry['COUNTRY (DISPLAY)'] == 'Somalia', 'Africa' , inplace=True)
RegionCountry['REGION (DISPLAY)'].mask(RegionCountry['COUNTRY (DISPLAY)'] == 'Sudan', 'Africa' , inplace=True)
RegionCountry['REGION (DISPLAY)'].mask(RegionCountry['COUNTRY (DISPLAY)'] == 'Libya', 'Africa' , inplace=True)
# Pays dans Eastern Meditteranean placés en Asie
RegionCountry.loc[RegionCountry['REGION (DISPLAY)'] =='Eastern Mediterranean','REGION (DISPLAY)'] ='Asia'
# Pays dans South-East Asia placés en Asie
RegionCountry.loc[RegionCountry['REGION (DISPLAY)'] =='South-East Asia','REGION (DISPLAY)'] ='Asia'
# Pays dans Western Pacific placés en Océanie
RegionCountry['REGION (DISPLAY)'].mask(RegionCountry['COUNTRY (DISPLAY)'] == 'Australia','Oceania',inplace=True)
RegionCountry['REGION (DISPLAY)'].mask(RegionCountry['COUNTRY (DISPLAY)'] == 'Fiji', 'Oceania',inplace=True)
RegionCountry['REGION (DISPLAY)'].mask(RegionCountry['COUNTRY (DISPLAY)'] == 'Papua New Guinea','Oceania', inplace=True)
RegionCountry['REGION (DISPLAY)'].mask(RegionCountry['COUNTRY (DISPLAY)'] == 'Fiji', 'Oceania' , inplace=True)
RegionCountry['REGION (DISPLAY)'].mask(RegionCountry['COUNTRY (DISPLAY)'] == 'Solomon Islands', 'Oceania' , inplace=True)
RegionCountry['REGION (DISPLAY)'].mask(RegionCountry['COUNTRY (DISPLAY)'] == 'Marshall Islands', 'Oceania' , inplace=True)
RegionCountry['REGION (DISPLAY)'].mask(RegionCountry['COUNTRY (DISPLAY)'] == 'Micronesia (Federated States of)', 'Oceania' , inplace=True)
RegionCountry['REGION (DISPLAY)'].mask(RegionCountry['COUNTRY (DISPLAY)'] == 'New Zealand', 'Oceania' , inplace=True)
RegionCountry['REGION (DISPLAY)'].mask(RegionCountry['COUNTRY (DISPLAY)'] == 'Vanuatu', 'Oceania' , inplace=True)
RegionCountry['REGION (DISPLAY)'].mask(RegionCountry['COUNTRY (DISPLAY)'] == 'Samoa', 'Oceania' , inplace=True)
RegionCountry['REGION (DISPLAY)'].mask(RegionCountry['COUNTRY (DISPLAY)'] == 'Palau', 'Oceania' , inplace=True)
RegionCountry['REGION (DISPLAY)'].mask(RegionCountry['COUNTRY (DISPLAY)'] == 'Kiribati', 'Oceania' , inplace=True)
RegionCountry['REGION (DISPLAY)'].mask(RegionCountry['COUNTRY (DISPLAY)'] == 'Tuvalu', 'Oceania' , inplace=True)
RegionCountry['REGION (DISPLAY)'].mask(RegionCountry['COUNTRY (DISPLAY)'] == 'Nauru', 'Oceania' , inplace=True)
RegionCountry['REGION (DISPLAY)'].mask(RegionCountry['COUNTRY (DISPLAY)'] == 'Tonga', 'Oceania' , inplace=True)
RegionCountry['REGION (DISPLAY)'].mask(RegionCountry['COUNTRY (DISPLAY)'] == 'Cook Islands', 'Oceania' , inplace=True)
RegionCountry['REGION (DISPLAY)'].mask(RegionCountry['COUNTRY (DISPLAY)'] == 'Niue', 'Oceania' , inplace=True)
# Pays dans Western Pacific placés en Asie
RegionCountry.loc[RegionCountry['REGION (DISPLAY)'] =='Western Pacific','REGION (DISPLAY)'] ='Asia'

In [338]:
RegionCountry[RegionCountry['REGION (DISPLAY)']=='Asia']

REGION (DISPLAY)                      COUNTRY (DISPLAY)
6               Asia                                Bahrain
7               Asia                             Bangladesh
10              Asia                                 Bhutan
15              Asia                               Cambodia
18              Asia                                  China
26              Asia  Democratic People's Republic of Korea
45              Asia                                  India
46              Asia             Iran (Islamic Republic of)
47              Asia                                   Iraq
51              Asia                                  Japan
52              Asia                                 Jordan
55              Asia                                 Kuwait
57              Asia       Lao People's Democratic Republic
59              Asia                                Lebanon
62              Asia                               Malaysia
66              Asia                               Mongolia
69              Asia                                  Nepal
77              Asia                                   Oman
78              Asia                               Pakistan
81              Asia                            Philippines
84              Asia                                  Qatar
85              Asia                      Republic of Korea
94              Asia                              Singapore
99              Asia                              Sri Lanka
115             Asia                            Afghanistan
125             Asia                      Brunei Darussalam
136             Asia                               Djibouti
148             Asia                              Indonesia
156             Asia                               Maldives
163             Asia                                Myanmar
174             Asia                           Saudi Arabia
182             Asia                   Syrian Arab Republic
183             Asia                               Thailand
184             Asia                            Timor-Leste
189             Asia                   United Arab Emirates
191             Asia                               Viet Nam
192             Asia                                  Yemen

In [339]:
RegionCountry['REGION (DISPLAY)'].value_counts()

Africa      53
Europe      53
Asia        37
Americas    35
Oceania     16
Name: REGION (DISPLAY), dtype: int64

# <a name="C4"> Correspondance WaterServices & Mortality </a>

Fusion des datasets sur le pays :

In [340]:
df = BasicAndSafelyManagedDrinkingWaterServices.merge(MortalityRateAttributedToWater,
                                                      how='outer',
                                                      on='Country',
                                                      indicator=True)

Décomptes des lignes sans correspondance avec mention du dataset d'origine :

In [341]:
df._merge.value_counts()

both          9882
left_only      594
right_only       0
Name: _merge, dtype: int64

Tous les pays de Mortality ont une correspondance dans WaterServices.

Affichage des pays dans WaterServices sans correspondance dans Mortality :

In [342]:
left_only = df[df._merge == 'left_only']
left_only['Country'].value_counts()

Nauru                    54
Cook Islands             54
San Marino               54
Monaco                   54
Tuvalu                   54
Marshall Islands         54
Niue                     54
Dominica                 54
Palau                    54
Saint Kitts and Nevis    54
Andorra                  54
Name: Country, dtype: int64

On ne dispose de données sur la mortalité pour ces pays.

Suppression de la colonne _merge :

In [343]:
df.drop(columns=['_merge'],inplace=True)

# <a name="C5"> Correspondance df & PoliticalStability </a>

Fusion des datasets sur le pays :

In [344]:
df = df.merge(PoliticalStability,
             how='outer',
             on='Country',
             indicator=True)

Décomptes des lignes sans correspondance :

In [345]:
df._merge.value_counts()

both          181278
left_only        216
right_only       169
Name: _merge, dtype: int64

Pays dans df sans correspondance dans politicalStability :

In [346]:
left_only = df[df._merge == 'left_only']
left_only['Country'].value_counts()

San Marino                     54
Republic of North Macedonia    54
Monaco                         54
China                          54
Name: Country, dtype: int64

Pays dans PoliticalStability sans correspondance dans df :

In [347]:
right_only = df[df._merge == 'right_only']
right_only['Country'].value_counts()

Palestine                    18
Bermuda                      18
China, Taiwan Province of    18
North Macedonia              18
China, Macao SAR             18
Puerto Rico                  18
China, Hong Kong SAR         18
China, mainland              18
American Samoa               15
Greenland                    10
Name: Country, dtype: int64

## Monaco, San Marino

Pas de données sur la stabilité politique.

## Macédoine 

changement du nom dans PoliticalStability pour établir la correspondance :

In [348]:
PoliticalStability.replace({'North Macedonia': 'Republic of North Macedonia'}, inplace=True)

Vérification :

In [349]:
PoliticalStability[PoliticalStability['Country'] == 'Republic of North Macedonia'].head()

Country  Year  Political_Stability Granularity
2320  Republic of North Macedonia  2000                -0.62       Total
2321  Republic of North Macedonia  2002                -1.04       Total
2322  Republic of North Macedonia  2003                -0.99       Total
2323  Republic of North Macedonia  2004                -0.85       Total
2324  Republic of North Macedonia  2005                -1.16       Total

## Puerto Rico, Bermudes, American Samoa, Palestine & Groënland

Aires géographiques avec un indice de stabilité politique qui ne sont pas recensés dans les dataset donnant les indicateurs sur l'eau.

Suppression des indicateurs de stabilité politique car ces aires géographiques prendront l'indicateur de stabilité politique de leurs pays d'appartenance.

In [350]:
PoliticalStability.drop(PoliticalStability[PoliticalStability['Country'] == 'Puerto Rico'].index,inplace=True)
PoliticalStability.drop(PoliticalStability[PoliticalStability['Country'] == 'Bermuda'].index,inplace=True)
PoliticalStability.drop(PoliticalStability[PoliticalStability['Country'] == 'American Samoa'].index,inplace=True)
PoliticalStability.drop(PoliticalStability[PoliticalStability['Country'] == 'Greenland'].index,inplace=True)
PoliticalStability.drop(PoliticalStability[PoliticalStability['Country'] == 'Palestine'].index,inplace=True)

Vérifications :

In [351]:
PoliticalStability[PoliticalStability['Country'] == 'Puerto Rico'].head()

Empty DataFrame
Columns: [Country, Year, Political_Stability, Granularity]
Index: []

In [352]:
PoliticalStability[PoliticalStability['Country'] == 'Bermuda'].head()

Empty DataFrame
Columns: [Country, Year, Political_Stability, Granularity]
Index: []

In [353]:
PoliticalStability[PoliticalStability['Country'] == 'American Samoa'].head()

Empty DataFrame
Columns: [Country, Year, Political_Stability, Granularity]
Index: []

In [354]:
PoliticalStability[PoliticalStability['Country'] == 'Greenland'].head()

Empty DataFrame
Columns: [Country, Year, Political_Stability, Granularity]
Index: []

In [355]:
PoliticalStability[PoliticalStability['Country'] == 'Palestine'].head()

Empty DataFrame
Columns: [Country, Year, Political_Stability, Granularity]
Index: []

## Chine

Suppression de l'indicateur de stabilité politique pour Honk-Kong, Macao et Taïwan car on ne dispose pas de données spécifiques sur l'accès à l'eau pour ces régions chinoises alors que les indicateurs sur l'eau semblent concerner tout le territoire de la Chine ("China").

On est donc obligés, en contrepartie, d'utiliser l'indicateur de stabilité politique de la Chine continentale pour ces zones.

### Suppression pour Honk-Kong :

In [356]:
PoliticalStability.drop(PoliticalStability[PoliticalStability['Country'] == 'China, Hong Kong SAR'].index,inplace=True)

Vérification :

In [357]:
PoliticalStability[PoliticalStability['Country'] == 'China, Hong Kong SAR']

Empty DataFrame
Columns: [Country, Year, Political_Stability, Granularity]
Index: []

### Suppression pour Macao :

In [358]:
PoliticalStability.drop(PoliticalStability[PoliticalStability['Country'] == 'China, Macao SAR'].index,inplace=True)

Vérification :

In [359]:
PoliticalStability[PoliticalStability['Country'] == 'China, Macao SAR']

Empty DataFrame
Columns: [Country, Year, Political_Stability, Granularity]
Index: []

### Suppression pour Taïwan

In [360]:
PoliticalStability.drop(PoliticalStability[PoliticalStability['Country'] == 'China, Taiwan Province of'].index,inplace=True)

Vérification :

In [361]:
PoliticalStability[PoliticalStability['Country'] == 'China, Taiwan Province of']

Empty DataFrame
Columns: [Country, Year, Political_Stability, Granularity]
Index: []

### changement du nom 'China, mainland' dans PoliticalStability en 'China' pour établir la correspondance :

In [362]:
PoliticalStability.replace({'China, mainland': 'China'}, inplace=True)

In [363]:
PoliticalStability[PoliticalStability['Country'] == 'China']

Country  Year  Political_Stability Granularity
699   China  2000                -0.21       Total
700   China  2002                -0.33       Total
701   China  2003                -0.56       Total
702   China  2004                -0.39       Total
703   China  2005                -0.50       Total
704   China  2006                -0.54       Total
705   China  2007                -0.50       Total
706   China  2008                -0.49       Total
707   China  2009                -0.45       Total
708   China  2010                -0.66       Total
709   China  2011                -0.60       Total
710   China  2012                -0.54       Total
711   China  2013                -0.54       Total
712   China  2014                -0.52       Total
713   China  2015                -0.55       Total
714   China  2016                -0.50       Total
715   China  2017                -0.23       Total
716   China  2018                -0.26       Total

Suppression de la colonne merge :

In [364]:
df.drop(columns=['_merge'],inplace=True)

## Vérification des modifications

Fusion de WaterServices et de Mortality :

In [365]:
df = BasicAndSafelyManagedDrinkingWaterServices.merge(MortalityRateAttributedToWater,
                                                      how='outer',
                                                      on='Country',
                                                      indicator=True)

Suppression de la colonne merge :

In [366]:
df.drop(columns=['_merge'],inplace=True)

Fusion de df et de PoliticalStability :

In [367]:
df = df.merge(PoliticalStability,
              how='outer',
              on='Country',
              indicator=True)

In [368]:
df['_merge'].value_counts()

both          183222
left_only        108
right_only         0
Name: _merge, dtype: int64

Pays dans df sans correspondance dans politicalStability :

In [369]:
left_only = df[df._merge == 'left_only']
left_only['Country'].value_counts()

San Marino    54
Monaco        54
Name: Country, dtype: int64

Vérification validée : il ne reste que les pays où les indicateurs de mortalité sont manquants.

Suppression de la colonne merge :

In [370]:
df.drop(columns=['_merge'],inplace=True)

In [371]:
len(df)

183330

# <a name="C6"> Correspondance df & Population </a>

Fusion des datasets sur le pays :

In [372]:
df = df.merge(Population,
             how='outer',
             on='Country',
             indicator=True)

Décomptes des lignes sans correspondance :

In [373]:
df._merge.value_counts()

both          10313730
right_only        2365
left_only          972
Name: _merge, dtype: int64

Pays dans df sans correspondance dans Population :

In [374]:
left_only = df[df._merge == 'left_only']
left_only['Country'].value_counts()

Republic of North Macedonia    972
Name: Country, dtype: int64

Pays dans Population sans correspondance dans df :

In [375]:
right_only = df[df._merge == 'right_only']
right_only = right_only['Country'].value_counts()
right_only

Greenland                                       57
Montserrat                                      57
China, mainland                                 57
Guam                                            57
Turks and Caicos Islands                        57
United States Virgin Islands                    57
Faroe Islands                                   57
French Guyana                                   57
Anguilla                                        57
Isle of Man                                     57
French Polynesia                                57
Guadeloupe                                      57
Palestine                                       57
China, Macao SAR                                57
Liechtenstein                                   57
Saint Pierre and Miquelon                       57
British Virgin Islands                          57
Netherlands Antilles (former)                   57
Cayman Islands                                  57
American Samoa                 

## Macédoine

Changement du nom 'North Macedonia" en "Republic of North Macedonia" dans Population :

In [376]:
Population.replace({'North Macedonia': 'Republic of North Macedonia'}, inplace=True)

Vérification :

In [377]:
Population[Population['Country'] == 'Republic of North Macedonia'].head()

Country Granularity  Year  Population
13904  Republic of North Macedonia       Total  2000    2034.819
13907  Republic of North Macedonia       Rural  2000     843.475
13908  Republic of North Macedonia       Urban  2000    1191.344
13909  Republic of North Macedonia       Total  2001    2042.842
13912  Republic of North Macedonia       Rural  2001     851.059

## Zones géographiques dans population

Ce sont des zones géographiques (pas des pays) dont on a indiqué la population.

La question est de savoir si, dans la population des pays de rattachement, la population de ces zones géographiques ont été ajoutées.

- Si oui : suppression des lignes dans Population
- Si non : Ajout de la popualtion de ces lignes aux pays de rattachement.

1er test sur la population de la Chine. Est-ce que l'entrée "China" prend en compte toute la population chinoise, avec Honk- Kong et Macao ?

In [378]:
Population[(Population['Country'] == 'China') & (Population['Year'] == 2018)]

Country Granularity  Year   Population
3876   China       Total  2018  1459377.612
3879   China       Rural  2018   583199.631
3880   China       Urban  2018   863601.691

1 459 377 612 habitants en 2018 selon l'entrée 'China' du datset Population.
- 1 milliard 427 millions selon Worldometer
- 1 milliard 403 millions selon la Banque Mondiale

Conclusion : l'entrée China semble prendre en compte les populations de Honk-Kong et de Macao.

2ème test sur la population du Royaume-Uni.

In [379]:
Population[(Population['Country'] == 'United Kingdom of Great Britain and Northern Ireland') \
           & (Population['Year'] == 2018)]

Country Granularity  Year  \
19712  United Kingdom of Great Britain and Northern I...       Total  2018   
19715  United Kingdom of Great Britain and Northern I...       Rural  2018   
19716  United Kingdom of Great Britain and Northern I...       Urban  2018   

       Population  
19712   67141.684  
19715   11052.278  
19716   55521.226

67 millions selon l'entrée Royaume-Uni du dataset Population.
- 66,46 millions selon la Banque mondiale.

Conclusion : l'entrée Royaume-Uni semble prendre en compte les aires géographiques mentionnées par ailleurs dans le dataset Population (Iles Vierges Britanniques par exemple).

Conclusion Générale : suppression des lignes du dataset Population sans correspondance dans dataset df.

## Suppression des lignes dans Population

In [380]:
type(right_only)

pandas.core.series.Series

Transformation de l'index de right_only en liste (contenant les noms des aires géographiques à supprimer) :

In [381]:
right_only = right_only.index.tolist()

Vérification :

In [382]:
type(right_only)

list

In [383]:
print(right_only)

['Greenland', 'Montserrat', 'China, mainland', 'Guam', 'Turks and Caicos Islands', 'United States Virgin Islands', 'Faroe Islands', 'French Guyana', 'Anguilla', 'Isle of Man', 'French Polynesia', 'Guadeloupe', 'Palestine', 'China, Macao SAR', 'Liechtenstein', 'Saint Pierre and Miquelon', 'British Virgin Islands', 'Netherlands Antilles (former)', 'Cayman Islands', 'American Samoa', 'Mayotte', 'Saint Helena, Ascension and Tristan da Cunha', 'Gibraltar', 'Falkland Islands (Malvinas)', 'Martinique', 'New Caledonia', 'Wallis and Futuna Islands', 'Holy See', 'Bermuda', 'Channel Islands', 'Aruba', 'Western Sahara', 'Puerto Rico', 'North Macedonia', 'Réunion', 'Northern Mariana Islands', 'China, Hong Kong SAR', 'China, Taiwan Province of', 'Tokelau', 'Sudan (former)', 'Curaçao', 'Sint Maarten  (Dutch part)', 'Bonaire, Sint Eustatius and Saba', 'Serbia and Montenegro', 'Saint Barthélemy', 'Saint-Martin (French part)']


Suppression des entrées dans Population :

In [384]:
for i in right_only :
    Population.drop(Population[Population['Country'] == i].index,inplace=True)

Vérification :

In [385]:
for i in right_only :
    print(Population[Population['Country'] == i])

Empty DataFrame
Columns: [Country, Granularity, Year, Population]
Index: []
Empty DataFrame
Columns: [Country, Granularity, Year, Population]
Index: []
Empty DataFrame
Columns: [Country, Granularity, Year, Population]
Index: []
Empty DataFrame
Columns: [Country, Granularity, Year, Population]
Index: []
Empty DataFrame
Columns: [Country, Granularity, Year, Population]
Index: []
Empty DataFrame
Columns: [Country, Granularity, Year, Population]
Index: []
Empty DataFrame
Columns: [Country, Granularity, Year, Population]
Index: []
Empty DataFrame
Columns: [Country, Granularity, Year, Population]
Index: []
Empty DataFrame
Columns: [Country, Granularity, Year, Population]
Index: []
Empty DataFrame
Columns: [Country, Granularity, Year, Population]
Index: []
Empty DataFrame
Columns: [Country, Granularity, Year, Population]
Index: []
Empty DataFrame
Columns: [Country, Granularity, Year, Population]
Index: []
Empty DataFrame
Columns: [Country, Granularity, Year, Population]
Index: []
Empty DataFr

Suppression de la colonne merge :

In [386]:
df.drop(columns=['_merge'],inplace=True)

## Vérification des modifications :

Fusion de WaterServices et de Mortality :

In [387]:
df = BasicAndSafelyManagedDrinkingWaterServices.merge(MortalityRateAttributedToWater,
                                                      how='outer',
                                                      on='Country',
                                                      indicator=True)

Suppression de la colonne merge :

In [388]:
df.drop(columns=['_merge'],inplace=True)

Fusion de df et de PoliticalStability :

In [389]:
df = df.merge(PoliticalStability,
              how='outer',
              on='Country',
              indicator=True)

In [390]:
df['_merge'].value_counts()

both          183222
left_only        108
right_only         0
Name: _merge, dtype: int64

Suppression de la colonne merge :

In [391]:
df.drop(columns=['_merge'],inplace=True)

Fusion de df et de Population :

In [392]:
df = df.merge(Population,
              how='outer',
              on='Country',
              indicator=True)

In [393]:
df['_merge'].value_counts()

both          10369134
right_only           0
left_only            0
Name: _merge, dtype: int64

Vérification validée.

Suppression de la colonne merge :

In [394]:
df.drop(columns=['_merge'],inplace=True)

# <a name="C7"> Correspondance df & RegionCountry </a>

In [395]:
RegionCountry.head()

REGION (DISPLAY) COUNTRY (DISPLAY)
0           Europe           Albania
1           Europe           Andorra
2           Europe           Armenia
3          Oceania         Australia
4           Europe           Austria

Il faut renommer la colonne 'COUNTRY (DISPLAY)' en 'Country' pour la fusion des données.

In [396]:
RegionCountry.rename(columns={"COUNTRY (DISPLAY)":"Country"}, inplace=True)

Vérification :

In [397]:
RegionCountry.head()

REGION (DISPLAY)    Country
0           Europe    Albania
1           Europe    Andorra
2           Europe    Armenia
3          Oceania  Australia
4           Europe    Austria

Fusion de df et de RegionCountry :

In [398]:
df = df.merge(RegionCountry,
              how='outer',
              on='Country',
              indicator=True)

In [399]:
df['_merge'].value_counts()

both          10369134
right_only           0
left_only            0
Name: _merge, dtype: int64

Toutes les entrées 'Country' correspondent.

# <a name="C8"> Fusion des datsets après nettoyage </a>

## <a name="C9"> Fusion WaterServices & Mortality </a>

Analyses des colonnes sur lesquelles fusionner :

In [400]:
BasicAndSafelyManagedDrinkingWaterServices.head()

Year      Country Granularity  \
0  2000  Afghanistan       Rural   
1  2000  Afghanistan       Total   
2  2000  Afghanistan       Urban   
3  2000      Albania       Rural   
4  2000      Albania       Total   

   Population using at least basic drinking-water services (%)  \
0                                           21.61913             
1                                           27.77190             
2                                           49.48745             
3                                           81.78472             
4                                           87.86662             

   Population using safely managed drinking-water services (%)  
0                                                NaN            
1                                                NaN            
2                                                NaN            
3                                                NaN            
4                                           49.29324

In [401]:
MortalityRateAttributedToWater.head()

Year              Country Granularity  \
2   2016          Afghanistan       Total   
5   2016              Albania       Total   
8   2016              Algeria       Total   
11  2016               Angola       Total   
14  2016  Antigua and Barbuda       Total   

    Mortality rate attributed to exposure to unsafe WASH services  WASH deaths  
2                                            13.92067               4824.35300  
5                                             0.16641                  4.86975  
8                                             1.86723                758.21000  
11                                           48.81467              14065.20000  
14                                            0.11403                  0.11513

In [402]:
temp_df = pd.pivot_table(MortalityRateAttributedToWater,
                         values=['WASH deaths'],
                         aggfunc='sum',
                         index='Granularity')
temp_df

WASH deaths
Granularity              
Total        870365.88011

Fusion sur les colonnes Country, year et Granularity :

In [403]:
df = BasicAndSafelyManagedDrinkingWaterServices.merge(MortalityRateAttributedToWater,
                                                      how='outer',
                                                      on=['Country','Year','Granularity'],
                                                      indicator=True)

Vérification de la fusion

In [404]:
df['_merge'].value_counts()

left_only     10293
both            183
right_only        0
Name: _merge, dtype: int64

183 lignes ont une correspondance. Nombre de lignes du dataset MortalityRateAttributedToWater :

In [405]:
len(MortalityRateAttributedToWater)

183

Toutes les données ont été fusionnées. Vérification de la fusion en filtrant sur l'année 2016, la granularité Total et l'indicateur _merge à 'both'.

On doit trouver 183 lignes pour être sur que ça ait été fusionné sur l'année 2016 et la granularité Total :

In [406]:
temp_df = df[(df['Year'] == 2016) & (df['Granularity'] == 'Total') & (df['_merge'] == 'both')]
print('Nombre de lignes :',len(temp_df))

Nombre de lignes : 183


In [407]:
temp_df

Year                                            Country Granularity  \
9313  2016                                        Afghanistan       Total   
9316  2016                                            Albania       Total   
9319  2016                                            Algeria       Total   
9325  2016                                             Angola       Total   
9328  2016                                Antigua and Barbuda       Total   
9331  2016                                          Argentina       Total   
9334  2016                                            Armenia       Total   
9337  2016                                          Australia       Total   
9340  2016                                            Austria       Total   
9343  2016                                         Azerbaijan       Total   
9346  2016                                            Bahamas       Total   
9349  2016                                            Bahrain       Total   
9352  2016                                         Bangladesh       Total   
9355  2016                                           Barbados       Total   
9358  2016                                            Belarus       Total   
9361  2016                                            Belgium       Total   
9364  2016                                             Belize       Total   
9367  2016                                              Benin       Total   
9370  2016                                             Bhutan       Total   
9373  2016                   Bolivia (Plurinational State of)       Total   
9376  2016                             Bosnia and Herzegovina       Total   
9379  2016                                           Botswana       Total   
9382  2016                                             Brazil       Total   
9385  2016                                  Brunei Darussalam       Total   
9388  2016                                           Bulgaria       Total   
9391  2016                                       Burkina Faso       Total   
9394  2016                                            Burundi       Total   
9397  2016                                         Cabo Verde       Total   
9400  2016                                           Cambodia       Total   
9403  2016                                           Cameroon       Total   
9406  2016                                             Canada       Total   
9409  2016                           Central African Republic       Total   
9412  2016                                               Chad       Total   
9415  2016                                              Chile       Total   
9418  2016                                              China       Total   
9421  2016                                           Colombia       Total   
9424  2016                                            Comoros       Total   
9427  2016                                              Congo       Total   
9433  2016                                         Costa Rica       Total   
9436  2016                                            Croatia       Total   
9439  2016                                               Cuba       Total   
9442  2016                                             Cyprus       Total   
9445  2016                                            Czechia       Total   
9448  2016                                      Côte d'Ivoire       Total   
9451  2016              Democratic People's Republic of Korea       Total   
9454  2016                   Democratic Republic of the Congo       Total   
9457  2016                                            Denmark       Total   
9460  2016                                           Djibouti       Total   
9466  2016                                 Dominican Republic       Total   
9469  2016                                            Ecuador       Total   
9472  2016                                              Egypt       Total   
94

Vérification validée.

Suppression de la colonne _merge

In [408]:
df.drop('_merge', inplace=True, axis=1)

Vérification de la suppression

In [409]:
df.head()

Year      Country Granularity  \
0  2000  Afghanistan       Rural   
1  2000  Afghanistan       Total   
2  2000  Afghanistan       Urban   
3  2000      Albania       Rural   
4  2000      Albania       Total   

   Population using at least basic drinking-water services (%)  \
0                                           21.61913             
1                                           27.77190             
2                                           49.48745             
3                                           81.78472             
4                                           87.86662             

   Population using safely managed drinking-water services (%)  \
0                                                NaN             
1                                                NaN             
2                                                NaN             
3                                                NaN             
4                                           49.29324             

   Mortality rate attributed to exposure to unsafe WASH services  WASH deaths  
0                                                NaN                      NaN  
1                                                NaN                      NaN  
2                                                NaN                      NaN  
3                                                NaN                      NaN  
4                                                NaN                      NaN

## <a name="C10"> Fusion df & Political Stability </a>

Analyse des colonnes de PoliticalStability pour savoir sur quelles colonnes fusionner :

In [410]:
PoliticalStability.head()

Country  Year  Political_Stability Granularity
0  Afghanistan  2000                -2.44       Total
1  Afghanistan  2002                -2.04       Total
2  Afghanistan  2003                -2.20       Total
3  Afghanistan  2004                -2.30       Total
4  Afghanistan  2005                -2.07       Total

Il faut fusionner sur Country, Year et Granularity.

Recherche des granularités présentes dans Political Stability :

In [411]:
PoliticalStability['Granularity'].value_counts()

Total    3393
Name: Granularity, dtype: int64

Il n'y a que la granularité Total.

Fusion des datasets :

In [412]:
df1 = df.merge(PoliticalStability,
              how='outer',
              on=['Country','Year','Granularity'],
              indicator=True)

Vérification de la fusion :

In [413]:
df1['_merge'].value_counts()

left_only     7273
both          3203
right_only     190
Name: _merge, dtype: int64

In [414]:
right_only = df1[df1['_merge'] == 'right_only']
right_only.head(20)

Year              Country Granularity  \
10476  2018          Afghanistan       Total   
10477  2018              Albania       Total   
10478  2018              Algeria       Total   
10479  2018              Andorra       Total   
10480  2018               Angola       Total   
10481  2018  Antigua and Barbuda       Total   
10482  2018            Argentina       Total   
10483  2018              Armenia       Total   
10484  2018            Australia       Total   
10485  2018              Austria       Total   
10486  2018           Azerbaijan       Total   
10487  2018              Bahamas       Total   
10488  2018              Bahrain       Total   
10489  2018           Bangladesh       Total   
10490  2018             Barbados       Total   
10491  2018              Belarus       Total   
10492  2018              Belgium       Total   
10493  2018               Belize       Total   
10494  2018                Benin       Total   
10495  2018               Bhutan       Total   

       Population using at least basic drinking-water services (%)  \
10476                                                NaN             
10477                                                NaN             
10478                                                NaN             
10479                                                NaN             
10480                                                NaN             
10481                                                NaN             
10482                                                NaN             
10483                                                NaN             
10484                                                NaN             
10485                                                NaN             
10486                                                NaN             
10487                                                NaN             
10488                                                NaN             
10489                                                NaN             
10490                                                NaN             
10491                                                NaN             
10492                                                NaN             
10493                                                NaN             
10494                                                NaN             
10495                                                NaN             

       Population using safely managed drinking-water services (%)  \
10476                                                NaN             
10477                                                NaN             
10478                                                NaN             
10479                                                NaN             
10480                                                NaN             
10481                                                NaN             
10482                                                NaN             
10483                                                NaN             
10484                                                NaN             
10485                                                NaN             
10486                                                NaN             
10487                                                NaN             
10488                                                NaN             
10489                                                NaN             
10490                                                NaN             
10491                                                NaN             
10492                                                NaN             
10493                                                NaN             
10494                                                NaN             
10495                                                NaN             

       Mortality rate attributed to exposure to unsafe W

In [415]:
right_only['Year'].value_counts()

2018    190
Name: Year, dtype: int64

Pas de fusion sur l'année 2018. Y'a-t-il une année 2018 dans df ?

In [416]:
df['Year'].value_counts().sort_index()

2000    582
2001    582
2002    582
2003    582
2004    582
2005    582
2006    582
2007    582
2008    582
2009    582
2010    582
2011    582
2012    582
2013    582
2014    582
2015    582
2016    582
2017    582
Name: Year, dtype: int64

Non, les right_only sont donc justifiés.

Il y a 3393 lignes dans le dataset PoliticalStability. 3203 lignes ont été fusionnées et 190 non (année 2018 manquante dans df). Toutes les lignes ont-elles été vérifiées ?

In [417]:
3203+190

3393

Oui.

Vérification de la fusion df-PoliticalStability validée.

In [418]:
df1.drop(columns=['_merge'],inplace=True)

## <a name="C11"> Fusion df & Population </a>

Lecture des colonnes de Population pour savoir sur quelles colonnes fusionner :

In [419]:
Population.head()

Country Granularity  Year  Population
0  Afghanistan       Total  2000   20779.953
3  Afghanistan       Rural  2000   15657.474
4  Afghanistan       Urban  2000    4436.282
5  Afghanistan       Total  2001   21606.988
8  Afghanistan       Rural  2001   16318.324

Il faut fusionner sur Country, Year et Granularity.

Recherche des granularités présentes dans Population :

In [420]:
Population['Granularity'].value_counts()

Urban    3650
Rural    3650
Total    3650
Name: Granularity, dtype: int64

Fusion des datasets :

In [421]:
df2 = df1.merge(Population,
              how='outer',
              on=['Country','Year','Granularity'],
              indicator=True)

Vérification de la fusion :

In [422]:
df2['_merge'].value_counts()

both          10558
right_only      392
left_only       108
Name: _merge, dtype: int64

In [423]:
right_only = df2[df2['_merge'] == 'right_only']
right_only.head()

Year      Country Granularity  \
10666  2018  Afghanistan       Rural   
10667  2018  Afghanistan       Urban   
10668  2018      Albania       Rural   
10669  2018      Albania       Urban   
10670  2018      Algeria       Rural   

       Population using at least basic drinking-water services (%)  \
10666                                                NaN             
10667                                                NaN             
10668                                                NaN             
10669                                                NaN             
10670                                                NaN             

       Population using safely managed drinking-water services (%)  \
10666                                                NaN             
10667                                                NaN             
10668                                                NaN             
10669                                                NaN             
10670                                                NaN             

       Mortality rate attributed to exposure to unsafe WASH services  \
10666                                                NaN               
10667                                                NaN               
10668                                                NaN               
10669                                                NaN               
10670                                                NaN               

       WASH deaths  Political_Stability  Population      _merge  
10666          NaN                  NaN   27099.874  right_only  
10667          NaN                  NaN    9273.302  right_only  
10668          NaN                  NaN    1164.388  right_only  
10669          NaN                  NaN    1769.975  right_only  
10670          NaN                  NaN   11498.038  right_only

Données de 2018 ?

In [424]:
right_only['Year'].value_counts()

2018    392
Name: Year, dtype: int64

Le dataset df1 ne contient, pour l'année 2018, que la granularité Total puisque :
- df (WaterServices + Mortality) n'a pas de données sur 2018
- PoliticalStability a des données sur 2018 sur la granularité Total

On s'attend donc à ne trouver que les granularités Rural et Urban sur right_only.

In [425]:
right_only['Granularity'].value_counts()

Rural    194
Urban    194
Total      4
Name: Granularity, dtype: int64

Affichage des granularités Total :

In [426]:
right_only[right_only['Granularity'] == 'Total']

Year       Country Granularity  \
10744  2018  Cook Islands       Total   
10891  2018        Monaco       Total   
10920  2018          Niue       Total   
10967  2018    San Marino       Total   

       Population using at least basic drinking-water services (%)  \
10744                                                NaN             
10891                                                NaN             
10920                                                NaN             
10967                                                NaN             

       Population using safely managed drinking-water services (%)  \
10744                                                NaN             
10891                                                NaN             
10920                                                NaN             
10967                                                NaN             

       Mortality rate attributed to exposure to unsafe WASH services  \
10744                                                NaN               
10891                                                NaN               
10920                                                NaN               
10967                                                NaN               

       WASH deaths  Political_Stability  Population      _merge  
10744          NaN                  NaN      17.518  right_only  
10891          NaN                  NaN      38.682  right_only  
10920          NaN                  NaN       1.620  right_only  
10967          NaN                  NaN      33.785  right_only

Hypothèse : les pays dans right_only n'existent pas dans PoliticalStability, au moins pour l'année 2018.

Vérification :

In [427]:
countries = right_only[right_only['Granularity'] == 'Total']
countries_list = countries['Country']
countries_list

10744    Cook Islands
10891          Monaco
10920            Niue
10967      San Marino
Name: Country, dtype: object

In [428]:
type(countries_list)

pandas.core.series.Series

In [429]:
for i in countries_list :
    print(PoliticalStability[PoliticalStability['Country'] == i])

          Country  Year  Political_Stability Granularity
789  Cook Islands  2009                 1.14       Total
790  Cook Islands  2010                 1.32       Total
Empty DataFrame
Columns: [Country, Year, Political_Stability, Granularity]
Index: []
     Country  Year  Political_Stability Granularity
2318    Niue  2009                 1.14       Total
2319    Niue  2010                 1.32       Total
Empty DataFrame
Columns: [Country, Year, Political_Stability, Granularity]
Index: []


Ces pays n'ont pas de données pour l'année 2018, ce qui explique qu'il n'y ait pas de fusion.

Vérification right_only validée.

Nombre de lignes de Population :

In [430]:
len(Population)

10950

Résultat de la fusion df1 + Population

In [431]:
df2['_merge'].value_counts()

both          10558
right_only      392
left_only       108
Name: _merge, dtype: int64

Si les lignes avec indicateurs both + right_only = 10950, toutes les lignes auront été vérifiées.

In [432]:
392+10558

10950

Fusion df1 + Population validée.

Suppression colonne _merge :

In [433]:
df2.drop(columns = '_merge', inplace=True)

## <a name="C12"> Fusion df2 & RegionCountry </a>

Affichage des colonnes de RegionCountry pour savoir sur quelles colonnes fusionner :

In [434]:
RegionCountry.head()

REGION (DISPLAY)    Country
0           Europe    Albania
1           Europe    Andorra
2           Europe    Armenia
3          Oceania  Australia
4           Europe    Austria

Fusion sur Country :

In [435]:
df3 = df2.merge(RegionCountry,
              how='outer',
              on='Country',
              indicator=True)

Vérification de la fusion :

In [436]:
df3['_merge'].value_counts()

both          11058
right_only        0
left_only         0
Name: _merge, dtype: int64

Fusion validée

In [437]:
df3.drop(columns = '_merge', inplace=True)

In [438]:
df3.head(50)

Year      Country Granularity  \
0   2000  Afghanistan       Rural   
1   2000  Afghanistan       Total   
2   2000  Afghanistan       Urban   
3   2001  Afghanistan       Rural   
4   2001  Afghanistan       Total   
5   2001  Afghanistan       Urban   
6   2002  Afghanistan       Rural   
7   2002  Afghanistan       Total   
8   2002  Afghanistan       Urban   
9   2003  Afghanistan       Rural   
10  2003  Afghanistan       Total   
11  2003  Afghanistan       Urban   
12  2004  Afghanistan       Rural   
13  2004  Afghanistan       Total   
14  2004  Afghanistan       Urban   
15  2005  Afghanistan       Rural   
16  2005  Afghanistan       Total   
17  2005  Afghanistan       Urban   
18  2006  Afghanistan       Rural   
19  2006  Afghanistan       Total   
20  2006  Afghanistan       Urban   
21  2007  Afghanistan       Rural   
22  2007  Afghanistan       Total   
23  2007  Afghanistan       Urban   
24  2008  Afghanistan       Rural   
25  2008  Afghanistan       Total   
26  2008  Afghanistan       Urban   
27  2009  Afghanistan       Rural   
28  2009  Afghanistan       Total   
29  2009  Afghanistan       Urban   
30  2010  Afghanistan       Rural   
31  2010  Afghanistan       Total   
32  2010  Afghanistan       Urban   
33  2011  Afghanistan       Rural   
34  2011  Afghanistan       Total   
35  2011  Afghanistan       Urban   
36  2012  Afghanistan       Rural   
37  2012  Afghanistan       Total   
38  2012  Afghanistan       Urban   
39  2013  Afghanistan       Rural   
40  2013  Afghanistan       Total   
41  2013  Afghanistan       Urban   
42  2014  Afghanistan       Rural   
43  2014  Afghanistan       Total   
44  2014  Afghanistan       Urban   
45  2015  Afghanistan       Rural   
46  2015  Afghanistan       Total   
47  2015  Afghanistan       Urban   
48  2016  Afghanistan       Rural   
49  2016  Afghanistan       Total   

    Population using at least basic drinking-water services (%)  \
0                                            21.61913             
1                                            27.77190             
2                                            49.48745             
3                                            21.61913             
4                                            27.79726             
5                                            49.48745             
6                                            23.59988             
7                                            29.90076             
8                                            51.90447             
9                                            25.58063             
10                                           32.00507             
11                                           54.32149             
12                                           27.56138             
13                                           34.12623             
14                                           56.73851             
15                                           29.54213             
16                                           36.26526             
17                                           59.15553             
18                                           31.52288             
19                                           38.40636             
20                                           61.57254             
21                                           33.73274             
22                                           40.84418             
23                                           64.50089             
24                                           35.96969             
25                                           43.31506             
26                                           67.46785             
27                                           38.23373             
28                                           45.81909             
29                                           70.47343             
30                                     

In [439]:
df3.to_csv('df.csv', index=False)

In [440]:
df4 = df3

# <a name="C14.2"> Ajout de la colonne Urban population rate </a>

Stockage granularité Total :

In [441]:
total = df4[df4['Granularity'] == 'Total']
total.head()

Year      Country Granularity  \
1   2000  Afghanistan       Total   
4   2001  Afghanistan       Total   
7   2002  Afghanistan       Total   
10  2003  Afghanistan       Total   
13  2004  Afghanistan       Total   

    Population using at least basic drinking-water services (%)  \
1                                            27.77190             
4                                            27.79726             
7                                            29.90076             
10                                           32.00507             
13                                           34.12623             

    Population using safely managed drinking-water services (%)  \
1                                                 NaN             
4                                                 NaN             
7                                                 NaN             
10                                                NaN             
13                                                NaN             

    Mortality rate attributed to exposure to unsafe WASH services  \
1                                                 NaN               
4                                                 NaN               
7                                                 NaN               
10                                                NaN               
13                                                NaN               

    WASH deaths  Political_Stability  Population REGION (DISPLAY)  
1           NaN                -2.44   20779.953             Asia  
4           NaN                  NaN   21606.988             Asia  
7           NaN                -2.04   22600.770             Asia  
10          NaN                -2.20   23680.871             Asia  
13          NaN                -2.30   24726.684             Asia

Stockage granularité Urban :

In [442]:
urban = df4[df4['Granularity'] == 'Urban']
urban.head()

Year      Country Granularity  \
2   2000  Afghanistan       Urban   
5   2001  Afghanistan       Urban   
8   2002  Afghanistan       Urban   
11  2003  Afghanistan       Urban   
14  2004  Afghanistan       Urban   

    Population using at least basic drinking-water services (%)  \
2                                            49.48745             
5                                            49.48745             
8                                            51.90447             
11                                           54.32149             
14                                           56.73851             

    Population using safely managed drinking-water services (%)  \
2                                                 NaN             
5                                                 NaN             
8                                                 NaN             
11                                                NaN             
14                                                NaN             

    Mortality rate attributed to exposure to unsafe WASH services  \
2                                                 NaN               
5                                                 NaN               
8                                                 NaN               
11                                                NaN               
14                                                NaN               

    WASH deaths  Political_Stability  Population REGION (DISPLAY)  
2           NaN                  NaN    4436.282             Asia  
5           NaN                  NaN    4648.139             Asia  
8           NaN                  NaN    4893.013             Asia  
11          NaN                  NaN    5155.788             Asia  
14          NaN                  NaN    5426.872             Asia

In [443]:
df5 = df4.merge(urban[['Country','Year','Population']],
                  how='left',
                  on=['Country','Year'],
                  indicator=True)
df5['_merge'].value_counts()

both          11058
right_only        0
left_only         0
Name: _merge, dtype: int64

In [444]:
df5.head()

Year      Country Granularity  \
0  2000  Afghanistan       Rural   
1  2000  Afghanistan       Total   
2  2000  Afghanistan       Urban   
3  2001  Afghanistan       Rural   
4  2001  Afghanistan       Total   

   Population using at least basic drinking-water services (%)  \
0                                           21.61913             
1                                           27.77190             
2                                           49.48745             
3                                           21.61913             
4                                           27.79726             

   Population using safely managed drinking-water services (%)  \
0                                                NaN             
1                                                NaN             
2                                                NaN             
3                                                NaN             
4                                                NaN             

   Mortality rate attributed to exposure to unsafe WASH services  WASH deaths  \
0                                                NaN                      NaN   
1                                                NaN                      NaN   
2                                                NaN                      NaN   
3                                                NaN                      NaN   
4                                                NaN                      NaN   

   Political_Stability  Population_x REGION (DISPLAY)  Population_y _merge  
0                  NaN     15657.474             Asia      4436.282   both  
1                -2.44     20779.953             Asia      4436.282   both  
2                  NaN      4436.282             Asia      4436.282   both  
3                  NaN     16318.324             Asia      4648.139   both  
4                  NaN     21606.988             Asia      4648.139   both

In [445]:
df5.drop(columns='_merge',inplace=True)

In [446]:
df5.head()

Year      Country Granularity  \
0  2000  Afghanistan       Rural   
1  2000  Afghanistan       Total   
2  2000  Afghanistan       Urban   
3  2001  Afghanistan       Rural   
4  2001  Afghanistan       Total   

   Population using at least basic drinking-water services (%)  \
0                                           21.61913             
1                                           27.77190             
2                                           49.48745             
3                                           21.61913             
4                                           27.79726             

   Population using safely managed drinking-water services (%)  \
0                                                NaN             
1                                                NaN             
2                                                NaN             
3                                                NaN             
4                                                NaN             

   Mortality rate attributed to exposure to unsafe WASH services  WASH deaths  \
0                                                NaN                      NaN   
1                                                NaN                      NaN   
2                                                NaN                      NaN   
3                                                NaN                      NaN   
4                                                NaN                      NaN   

   Political_Stability  Population_x REGION (DISPLAY)  Population_y  
0                  NaN     15657.474             Asia      4436.282  
1                -2.44     20779.953             Asia      4436.282  
2                  NaN      4436.282             Asia      4436.282  
3                  NaN     16318.324             Asia      4648.139  
4                  NaN     21606.988             Asia      4648.139

In [447]:
df5.rename(columns={"Population_x": "Population","Population_y": "Urban population"},inplace=True)

In [448]:
df5.head()

Year      Country Granularity  \
0  2000  Afghanistan       Rural   
1  2000  Afghanistan       Total   
2  2000  Afghanistan       Urban   
3  2001  Afghanistan       Rural   
4  2001  Afghanistan       Total   

   Population using at least basic drinking-water services (%)  \
0                                           21.61913             
1                                           27.77190             
2                                           49.48745             
3                                           21.61913             
4                                           27.79726             

   Population using safely managed drinking-water services (%)  \
0                                                NaN             
1                                                NaN             
2                                                NaN             
3                                                NaN             
4                                                NaN             

   Mortality rate attributed to exposure to unsafe WASH services  WASH deaths  \
0                                                NaN                      NaN   
1                                                NaN                      NaN   
2                                                NaN                      NaN   
3                                                NaN                      NaN   
4                                                NaN                      NaN   

   Political_Stability  Population REGION (DISPLAY)  Urban population  
0                  NaN   15657.474             Asia          4436.282  
1                -2.44   20779.953             Asia          4436.282  
2                  NaN    4436.282             Asia          4436.282  
3                  NaN   16318.324             Asia          4648.139  
4                  NaN   21606.988             Asia          4648.139

In [449]:
df5.loc[df5['Granularity'] == 'Total', 'Urban population rate'] = \
(df5['Urban population']/df5['Population'])*100

In [450]:
df5.head()

Year      Country Granularity  \
0  2000  Afghanistan       Rural   
1  2000  Afghanistan       Total   
2  2000  Afghanistan       Urban   
3  2001  Afghanistan       Rural   
4  2001  Afghanistan       Total   

   Population using at least basic drinking-water services (%)  \
0                                           21.61913             
1                                           27.77190             
2                                           49.48745             
3                                           21.61913             
4                                           27.79726             

   Population using safely managed drinking-water services (%)  \
0                                                NaN             
1                                                NaN             
2                                                NaN             
3                                                NaN             
4                                                NaN             

   Mortality rate attributed to exposure to unsafe WASH services  WASH deaths  \
0                                                NaN                      NaN   
1                                                NaN                      NaN   
2                                                NaN                      NaN   
3                                                NaN                      NaN   
4                                                NaN                      NaN   

   Political_Stability  Population REGION (DISPLAY)  Urban population  \
0                  NaN   15657.474             Asia          4436.282   
1                -2.44   20779.953             Asia          4436.282   
2                  NaN    4436.282             Asia          4436.282   
3                  NaN   16318.324             Asia          4648.139   
4                  NaN   21606.988             Asia          4648.139   

   Urban population rate  
0                    NaN  
1              21.348855  
2                    NaN  
3                    NaN  
4              21.512202

# <a name="C14.5"> Ajout des densités de Population </a>

## <a name="C14.51"> Ajout des densités de Population par Pays</a>

In [451]:
pop_density_country = pd.read_excel('population density Country calculated from World Bank.xlsx', index_col=0)
pop_density_country.head()

Country Granularity  Population density
Year                                             
2000  Afghanistan       Total           31.829117
2001  Afghanistan       Total           33.095904
2002  Afghanistan       Total           34.618102
2003  Afghanistan       Total           36.272510
2004  Afghanistan       Total           37.874413

In [452]:
df6 = df5.merge(pop_density_country,
                how='left',
                on=['Country','Year','Granularity'],
                indicator=True)

In [453]:
df6['_merge'].value_counts()

left_only     7421
both          3637
right_only       0
Name: _merge, dtype: int64

Les granularités Urban et Rural n'ont pas pu être jointées.

In [454]:
Urban = df6[df6['Granularity'] == 'Urban']
Rural = df6[df6['Granularity'] == 'Rural']
len(Urban)+len(Rural)

7372

Sur les 7 421 'left_only', 7 372 lignes justifiées par une granularité Rural ou Urban.

Combien de 'left_only' reste-t-il à justifier ?

In [455]:
7421-7372

49

In [456]:
len(df6[(df6['Granularity'] == 'Total') & (df6['_merge'] == 'left_only')])

49

La sélection comprend bien les 49 lignes restant à justifier.

In [457]:
df6[(df6['Granularity'] == 'Total') & (df6['_merge'] == 'left_only')]

Year       Country Granularity  \
2224  2000  Cook Islands       Total   
2227  2001  Cook Islands       Total   
2230  2002  Cook Islands       Total   
2233  2003  Cook Islands       Total   
2236  2004  Cook Islands       Total   
2239  2005  Cook Islands       Total   
2242  2006  Cook Islands       Total   
2245  2007  Cook Islands       Total   
2248  2008  Cook Islands       Total   
2251  2009  Cook Islands       Total   
2254  2010  Cook Islands       Total   
2257  2011  Cook Islands       Total   
2260  2012  Cook Islands       Total   
2263  2013  Cook Islands       Total   
2266  2014  Cook Islands       Total   
2269  2015  Cook Islands       Total   
2272  2016  Cook Islands       Total   
2275  2017  Cook Islands       Total   
2277  2018  Cook Islands       Total   
7183  2000          Niue       Total   
7186  2001          Niue       Total   
7189  2002          Niue       Total   
7192  2003          Niue       Total   
7195  2004          Niue       Total   
7198  2005          Niue       Total   
7201  2006          Niue       Total   
7204  2007          Niue       Total   
7207  2008          Niue       Total   
7210  2009          Niue       Total   
7213  2010          Niue       Total   
7216  2011          Niue       Total   
7219  2012          Niue       Total   
7222  2013          Niue       Total   
7225  2014          Niue       Total   
7228  2015          Niue       Total   
7231  2016          Niue       Total   
7234  2017          Niue       Total   
7236  2018          Niue       Total   
9235  2000   South Sudan       Total   
9238  2001   South Sudan       Total   
9241  2002   South Sudan       Total   
9244  2003   South Sudan       Total   
9247  2004   South Sudan       Total   
9250  2005   South Sudan       Total   
9253  2006   South Sudan       Total   
9256  2007   South Sudan       Total   
9259  2008   South Sudan       Total   
9262  2009   South Sudan       Total   
9265  2010   South Sudan       Total   

      Population using at least basic drinking-water services (%)  \
2224                                           99.89700             
2227                                           99.89565             
2230                                           99.89432             
2233                                           99.89298             
2236                                           99.89164             
2239                                           99.89030             
2242                                           99.88897             
2245                                           99.88763             
2248                                           99.88629             
2251                                           99.88495             
2254                                           99.88361             
2257                                           99.88228             
2260                                           99.88094             
2263                                           99.87960             
2266                                           99.87960             
2269                                           99.87960             
2272                                           99.87960             
2275                                           99.87960             
2277                                                NaN             
7183                                           99.22893             
7186                                           99.22893             
7189                                           99.22893             
7192                                           99.22893             
7195                                           99.22893             
7198                                           99.11446             
7201                                           99.00000             
7204                                           98.88554             
7207                                           98.77107             
7210

Pas de données sur Cook Islands et Inue. Données parcellaires sur Soudan du Sud.

Fusion validée.

In [458]:
df6.drop(columns='_merge', inplace=True)

In [459]:
df6 = df6.rename(columns={'Population density':'Population density Country'})

In [460]:
df6['REGION (DISPLAY)'].value_counts()

Europe      3021
Africa      3021
Asia        2109
Americas    1995
Oceania      912
Name: REGION (DISPLAY), dtype: int64

## <a name="C14.52"> Ajout des densités de Population par Région</a>

Données provenant de https://www.populationdata.net/monde/

In [461]:
# create a list of our conditions
conditions = [
    (df6['REGION (DISPLAY)'] == 'Asia'),
    (df6['REGION (DISPLAY)'] == 'Africa'),
    (df6['REGION (DISPLAY)'] == 'Americas'),
    (df6['REGION (DISPLAY)'] == 'Oceania'),
    (df6['REGION (DISPLAY)'] == 'Europe'),
    ]

# create a list of the values we want to assign for each condition
values = [142.7, 43.25, 24.29, 5.06, 32.32]

# create a new column and use np.select to assign values to it using our lists as arguments
df6['population density region'] = np.select(conditions, values)

# display updated DataFrame
df6['population density region'].value_counts()

43.25     3021
32.32     3021
142.70    2109
24.29     1995
5.06       912
Name: population density region, dtype: int64

In [462]:
df6.head()

Year      Country Granularity  \
0  2000  Afghanistan       Rural   
1  2000  Afghanistan       Total   
2  2000  Afghanistan       Urban   
3  2001  Afghanistan       Rural   
4  2001  Afghanistan       Total   

   Population using at least basic drinking-water services (%)  \
0                                           21.61913             
1                                           27.77190             
2                                           49.48745             
3                                           21.61913             
4                                           27.79726             

   Population using safely managed drinking-water services (%)  \
0                                                NaN             
1                                                NaN             
2                                                NaN             
3                                                NaN             
4                                                NaN             

   Mortality rate attributed to exposure to unsafe WASH services  WASH deaths  \
0                                                NaN                      NaN   
1                                                NaN                      NaN   
2                                                NaN                      NaN   
3                                                NaN                      NaN   
4                                                NaN                      NaN   

   Political_Stability  Population REGION (DISPLAY)  Urban population  \
0                  NaN   15657.474             Asia          4436.282   
1                -2.44   20779.953             Asia          4436.282   
2                  NaN    4436.282             Asia          4436.282   
3                  NaN   16318.324             Asia          4648.139   
4                  NaN   21606.988             Asia          4648.139   

   Urban population rate  Population density Country  \
0                    NaN                         NaN   
1              21.348855                   31.829117   
2                    NaN                         NaN   
3                    NaN                         NaN   
4              21.512202                   33.095904   

   population density region  
0                      142.7  
1                      142.7  
2                      142.7  
3                      142.7  
4                      142.7

In [463]:
df7 = df6

## <a name="C14.53"> Ajout de la densité de Population au niveau Monde</a>

In [464]:
df7['population density world'] = 50.2

In [465]:
df7.head()

Year      Country Granularity  \
0  2000  Afghanistan       Rural   
1  2000  Afghanistan       Total   
2  2000  Afghanistan       Urban   
3  2001  Afghanistan       Rural   
4  2001  Afghanistan       Total   

   Population using at least basic drinking-water services (%)  \
0                                           21.61913             
1                                           27.77190             
2                                           49.48745             
3                                           21.61913             
4                                           27.79726             

   Population using safely managed drinking-water services (%)  \
0                                                NaN             
1                                                NaN             
2                                                NaN             
3                                                NaN             
4                                                NaN             

   Mortality rate attributed to exposure to unsafe WASH services  WASH deaths  \
0                                                NaN                      NaN   
1                                                NaN                      NaN   
2                                                NaN                      NaN   
3                                                NaN                      NaN   
4                                                NaN                      NaN   

   Political_Stability  Population REGION (DISPLAY)  Urban population  \
0                  NaN   15657.474             Asia          4436.282   
1                -2.44   20779.953             Asia          4436.282   
2                  NaN    4436.282             Asia          4436.282   
3                  NaN   16318.324             Asia          4648.139   
4                  NaN   21606.988             Asia          4648.139   

   Urban population rate  Population density Country  \
0                    NaN                         NaN   
1              21.348855                   31.829117   
2                    NaN                         NaN   
3                    NaN                         NaN   
4              21.512202                   33.095904   

   population density region  population density world  
0                      142.7                      50.2  
1                      142.7                      50.2  
2                      142.7                      50.2  
3                      142.7                      50.2  
4                      142.7                      50.2

# <a name="C15"> Export du .csv </a>

In [466]:
df7.to_csv('df.csv', index=False)

In [467]:
df7[df7['Country'] == 'China']

Year Country Granularity  \
1995  2000   China       Rural   
1996  2000   China       Total   
1997  2000   China       Urban   
1998  2001   China       Rural   
1999  2001   China       Total   
2000  2001   China       Urban   
2001  2002   China       Rural   
2002  2002   China       Total   
2003  2002   China       Urban   
2004  2003   China       Rural   
2005  2003   China       Total   
2006  2003   China       Urban   
2007  2004   China       Rural   
2008  2004   China       Total   
2009  2004   China       Urban   
2010  2005   China       Rural   
2011  2005   China       Total   
2012  2005   China       Urban   
2013  2006   China       Rural   
2014  2006   China       Total   
2015  2006   China       Urban   
2016  2007   China       Rural   
2017  2007   China       Total   
2018  2007   China       Urban   
2019  2008   China       Rural   
2020  2008   China       Total   
2021  2008   China       Urban   
2022  2009   China       Rural   
2023  2009   China       Total   
2024  2009   China       Urban   
2025  2010   China       Rural   
2026  2010   China       Total   
2027  2010   China       Urban   
2028  2011   China       Rural   
2029  2011   China       Total   
2030  2011   China       Urban   
2031  2012   China       Rural   
2032  2012   China       Total   
2033  2012   China       Urban   
2034  2013   China       Rural   
2035  2013   China       Total   
2036  2013   China       Urban   
2037  2014   China       Rural   
2038  2014   China       Total   
2039  2014   China       Urban   
2040  2015   China       Rural   
2041  2015   China       Total   
2042  2015   China       Urban   
2043  2016   China       Rural   
2044  2016   China       Total   
2045  2016   China       Urban   
2046  2017   China       Rural   
2047  2017   China       Total   
2048  2017   China       Urban   
2049  2018   China       Total   
2050  2018   China       Rural   
2051  2018   China       Urban   

      Population using at least basic drinking-water services (%)  \
1995                                           70.40273             
1996                                           80.39258             
1997                                           98.24744             
1998                                           70.40273             
1999                                           80.73117             
2000                                           98.24744             
2001                                           71.40005             
2002                                           81.70317             
2003                                           98.21362             
2004                                           72.39563             
2005                                           82.65153             
2006                                           98.17977             
2007                                           73.38948             
2008                                           83.57525             
2009                                           98.14588             
2010                                           74.38159             
2011                                           84.47221             
2012                                           98.11195             
2013                                           75.37196             
2014                                           85.33264             
2015                                           98.07798             
2016                                           76.36061             
2017                                           86.16128             
2018                                           98.04398             
2019                                           77.34751             
2020                                           86.96360             
2021                                           98.00993             
2022                                           78.33269             
2023                            

In [468]:
df7.head(50)

Year      Country Granularity  \
0   2000  Afghanistan       Rural   
1   2000  Afghanistan       Total   
2   2000  Afghanistan       Urban   
3   2001  Afghanistan       Rural   
4   2001  Afghanistan       Total   
5   2001  Afghanistan       Urban   
6   2002  Afghanistan       Rural   
7   2002  Afghanistan       Total   
8   2002  Afghanistan       Urban   
9   2003  Afghanistan       Rural   
10  2003  Afghanistan       Total   
11  2003  Afghanistan       Urban   
12  2004  Afghanistan       Rural   
13  2004  Afghanistan       Total   
14  2004  Afghanistan       Urban   
15  2005  Afghanistan       Rural   
16  2005  Afghanistan       Total   
17  2005  Afghanistan       Urban   
18  2006  Afghanistan       Rural   
19  2006  Afghanistan       Total   
20  2006  Afghanistan       Urban   
21  2007  Afghanistan       Rural   
22  2007  Afghanistan       Total   
23  2007  Afghanistan       Urban   
24  2008  Afghanistan       Rural   
25  2008  Afghanistan       Total   
26  2008  Afghanistan       Urban   
27  2009  Afghanistan       Rural   
28  2009  Afghanistan       Total   
29  2009  Afghanistan       Urban   
30  2010  Afghanistan       Rural   
31  2010  Afghanistan       Total   
32  2010  Afghanistan       Urban   
33  2011  Afghanistan       Rural   
34  2011  Afghanistan       Total   
35  2011  Afghanistan       Urban   
36  2012  Afghanistan       Rural   
37  2012  Afghanistan       Total   
38  2012  Afghanistan       Urban   
39  2013  Afghanistan       Rural   
40  2013  Afghanistan       Total   
41  2013  Afghanistan       Urban   
42  2014  Afghanistan       Rural   
43  2014  Afghanistan       Total   
44  2014  Afghanistan       Urban   
45  2015  Afghanistan       Rural   
46  2015  Afghanistan       Total   
47  2015  Afghanistan       Urban   
48  2016  Afghanistan       Rural   
49  2016  Afghanistan       Total   

    Population using at least basic drinking-water services (%)  \
0                                            21.61913             
1                                            27.77190             
2                                            49.48745             
3                                            21.61913             
4                                            27.79726             
5                                            49.48745             
6                                            23.59988             
7                                            29.90076             
8                                            51.90447             
9                                            25.58063             
10                                           32.00507             
11                                           54.32149             
12                                           27.56138             
13                                           34.12623             
14                                           56.73851             
15                                           29.54213             
16                                           36.26526             
17                                           59.15553             
18                                           31.52288             
19                                           38.40636             
20                                           61.57254             
21                                           33.73274             
22                                           40.84418             
23                                           64.50089             
24                                           35.96969             
25                                           43.31506             
26                                           67.46785             
27                                           38.23373             
28                                           45.81909             
29                                           70.47343             
30                                     